### PR TITLE
finer-tuned handling of ambiguous XML nodes for infix operators

### PR DESCRIPTION
**Pull request by MichaelChirico**
Towards #1977. This is _sort of_ a breaking change, though it doesn't affect any defaults, but I think the new behavior is much cleaner.

### CODE REVIEW

File: NEWS.md [hunk 1]
```diff
@@ -52,6 +52,7 @@
   + assumes `glue()` is `glue::glue()` when `interpret_glue=TRUE` (#2032, @MichaelChirico).
   + finds function usages, including infix usage, inside `glue()` calls to avoid false positives for "unused objects" (#2029 and #2069, @MichaelChirico).
 * `object_name_linter()` no longer attempts to lint strings in function calls on the LHS of assignments (#1466, @MichaelChirico).
+* `infix_spaces_linter()` distinguishes `<-`, `:=`, `<<-` and `->`, `->>`, i.e. `infix_spaces_linter(exclude_operator = "->")` will no longer exclude `->>` (#2115, @MichaelChirico).
```
Comment (AshesITR) [inline NEWS.md:55]: This should be put under a ## Breaking Changes section

File: R/shared_constants.R [hunk 2]
```diff
@@ -168,17 +172,14 @@ infix_metadata$unary <- infix_metadata$xml_tag %in% c("OP-PLUS", "OP-MINUS", "OP
 # high-precedence operators are ignored by this linter; see
 #   https://style.tidyverse.org/syntax.html#infix-operators
 infix_metadata$low_precedence <- infix_metadata$string_value %in% c(
```
Comment (AshesITR) [inline R/shared_constants.R:174]: WDYT about moving infix_metadata to a csv file using sysdata.rda?

Comment (MichaelChirico) [reply R/shared_constants.R:174]: my sense is it would bring more trouble than it's worth

File: R/undesirable_operator_linter.R [hunk 3]
```diff
@@ -46,23 +46,18 @@ undesirable_operator_linter <- function(op = default_undesirable_operators) {
...
+  ]
 
-  included_operators <- undesirable_operator_metadata$string_value %in% names(op) |
-    undesirable_operator_metadata$exact_string_value %in% names(op)
+  included_operators <- undesirable_operator_metadata$string_value %in% names(op)
   operator_nodes <- undesirable_operator_metadata$xml_tag[included_operators]
-  needs_exact_string <- !is.na(undesirable_operator_metadata$exact_string_value[included_operators])
+  needs_exact_string <- operator_nodes %in% infix_metadata$xml_tag[infix_metadata$ambiguous_node]
   operator_nodes[needs_exact_string] <- sprintf(
     "%s[text() = '%s']",
```
Comment (AshesITR) [inline R/undesirable_operator_linter.R:58]: This snippet could be refactored into infix_metadata (`xml_token`?) to avoid duplication.

**Comment by MichaelChirico** [comment]
FWIW this does not affect anyone on GitHub:

https://github.com/search?q=infix_spaces_linter+exclude_operators+-repo%3Ar-lib%2Flintr+-is%3Afork+-repo%3Acran%2Flintr&type=code

**Review by AshesITR — APPROVED** [review]

**Pull request merged**